src/blueprint.zig

```diff
@@ -115,45 +115,33 @@ pub const Blueprint = struct {
 	}
 
 	fn _pasteInGeneration(self: Blueprint, pos: Vec3i, chunk: *ServerChunk, comptime mode: PasteMode) void {
-		const chunkWidth: i32 = @intCast(chunk.super.width);
-
-		const blueprintWidth: i32 = @intCast(self.blocks.width);
-		const blueprintDepth: i32 = @intCast(self.blocks.depth);
-		const blueprintHeight: i32 = @intCast(self.blocks.height);
-
-		var blueprintX: u32 = @max(0, -pos[0]);
-		const blueprintEndX = @min(blueprintWidth, chunkWidth - pos[0]);
-
-		var chunkX = @max(0, pos[0]);
-		const chunkEndX = @min(chunkWidth, blueprintWidth + pos[0]);
-
-		while(chunkX < chunkEndX and blueprintX < blueprintEndX) : ({
-			chunkX += chunk.super.pos.voxelSize;
-			blueprintX += chunk.super.pos.voxelSize;
-		}) {
-			var blueprintY: u32 = @max(0, -pos[1]);
-			const blueprintEndY = @min(blueprintDepth, chunkWidth - pos[1]);
-
-			var chunkY = @max(0, pos[1]);
-			const chunkEndY = @min(chunkWidth, blueprintDepth + pos[1]);
-
-			while(chunkY < chunkEndY and blueprintY < blueprintEndY) : ({
-				chunkY += chunk.super.pos.voxelSize;
-				blueprintY += chunk.super.pos.voxelSize;
-			}) {
-				var blueprintZ: u32 = @max(0, -pos[2]);
-				const blueprintEndZ = @min(blueprintHeight, chunkWidth - pos[2]);
-
-				var chunkZ = @max(0, pos[2]);
-				const chunkEndZ = @min(chunkWidth, blueprintHeight + pos[2]);
-
-				while(chunkZ < chunkEndZ and blueprintZ < blueprintEndZ) : ({
-					chunkZ += chunk.super.pos.voxelSize;
-					blueprintZ += chunk.super.pos.voxelSize;
-				}) {
+		const blueprintOffset = @max(@as(Vec3i, @splat(0)), -pos);
+		const chunkOffset = @max(@as(Vec3i, @splat(0)), pos);
+
+		const indexEndX: i32 = @min(@as(i32, @intCast(chunk.super.width)) - chunkOffset[0], @as(i32, @intCast(self.blocks.width)) - blueprintOffset[0]);
```

Comment:
Something like this could work.
```suggestion
		const blueprintEndX: i32 = @min(chunk.super.width - pos[0], @as(i32, @intCast(self.blocks.width)));
```
Then you could iterate the blueprint coordinates directly, and the chunk-relative coordinates are just `blueprintX - pos[0]`

Reply:
pos is chunk local relative so it can be negative, cant be used directly as index for blueprint. Changed to iterate blueprint coordinates.